File: heos-track-listener/models/HeosPlayedTrack.js
const mongoose = require("mongoose");

const redis = require("redis").createClient({
  host: process.env.REDIS_HOST,
  port: process.env.REDIS_PORT,
});

const Schema = mongoose.Schema;

const heosPlayedTrackSchema = new Schema({
  type: String,
  source: String,
  title: String,
  artist: String,
  album: String,
  duration: Number,
  imageUrl: String,
  startedAt: Number,
  finishedAt: Number,
  ready: {
    nowPlaying: {
      type: Boolean,
      default: false,
    },
    track: {
      type: Boolean,
      default: false,
    },
  },
  submit: {
    nowPlaying: {
      type: Boolean,
      default: false,
    },
    track: {
      type: Boolean,
      default: false,
    },
  },
  submitStatus: {
    nowPlaying: {
      type: Boolean,
      default: false,
    },
    track: {
      type: Boolean,
      default: false,
    },
  },
  player: String,
});

heosPlayedTrackSchema.post("save", (track) => {
  if (track.ready.nowPlaying) {
    redis.publish(process.env.REDIS_CHANNEL, track._id.toString());
  }
});

module.exports = mongoose.model(
  "HeosPlayedTrack",
  heosPlayedTrackSchema,
  "heosPlayedTracks"
);

File: heos-track-scrobbler/models/heos_played_track.py
from mongoengine import *


class Submit(EmbeddedDocument):
    nowPlaying = BooleanField(default=False)
    track = BooleanField(default=False)


class HeosPlayedTrack(Document):
    type = StringField()
    source = StringField()
    title = StringField()
    artist = StringField()
    album = StringField()
    duration = FloatField()
    imageUrl = StringField()
    startedAt = IntField(min_value=0)
    finishedAt = IntField(min_value=0)
    ready = EmbeddedDocumentField(Submit)
    submit = EmbeddedDocumentField(Submit)
    submitStatus = EmbeddedDocumentField(Submit)
    player = StringField()
    v = IntField(db_field="__v")

    meta = {"collection": "heosPlayedTracks", "db_alias": "heosScrobbler"}
